nexus/src/db/subnet_allocation.rs

```diff
@@ -219,9 +224,278 @@ impl QueryFragment<Pg> for AllocateIpQueryValues {
     }
 }
 
+/// Generate a CTE that can be used to insert a VPC Subnet, only if the IP
+/// address ranges of that subnet don't overlap with existing Subnets in the
+/// same VPC.
+///
+/// In particular, this generates a CTE like so:
+///
+/// ```sql
+/// WITH candidate(
+///     id,
+///     name,
+///     description,
+///     time_created,
+///     time_modified,
+///     time_deleted,
+///     vpc_id,
+///     ipv4_block,
+///     ipv6_block
+/// ) AS (VALUES (
+///     <id>,
+///     <name>,
+///     <description>,
+///     <time_created>,
+///     <time_modified>,
+///     NULL::TIMESTAMPTZ,
+///     <vpc_id>,
+///     <ipv4_block>,
+///     <ipv6_block>
+/// ))
+/// SELECT *
+/// FROM candidate
+/// WHERE NOT EXISTS (
+///     SELECT ipv4_block, ipv6_block
+///     FROM vpc_subnet
+///     WHERE
+///         vpc_id = <vpc_id> AND
+///         time_deleted IS NULL AND
+///         (
+///             (ipv4_block && candidate.ipv4_block) OR
+///             (ipv6_block && candidate.ipv6_block)
+///         )
+/// )
+/// ```
+///
+/// This can be used
+///
+/// Notes
+/// -----
+///
+/// This might be better done with an exclusion constraint on the actual table,
```

Comment:
Is it possible to insert a conflicting subnet?  (Does the timing matter at all?)

Reply:
If you're asking, would it be possible with an exclusion constraint, I'm really not sure. I'd expect the answer is "no", but I'm guessing.

If you're asking is it possible now, with these changes, I think the answer is "probably". I'm embarrassed to admit it didn't even occur to me to put this in a transaction. I'm happy to do that, if you think it's the right way to go.

Reply:
@davepacheco and I spoke more about this. The take-home was that, since this is all part of a single query, there shouldn't be a way to get conflicts due to Cockroach's serializability guarantees.